lib/groups.js

```diff
@@ -0,0 +1,144 @@
+var ERR = require('async-stacktrace');
+
+var sqldb = require('../prairielib/lib/sql-db');
+var sqlLoader = require('../prairielib/lib/sql-loader');
+var sql = sqlLoader.loadSqlEquiv(__filename);
+
+module.exports.getGroupInfo = function (res, callback) {
+  var params = {
+    assessment_id: res.locals.assessment.id,
+    user_id: res.locals.user.user_id,
+  };
+
+  sqldb.query(sql.get_config_info, params, function (err, result) {
+    if (ERR(err, callback)) return;
+    res.locals.permissions = result.rows[0];
+    res.locals.minsize = result.rows[0].minimum || 0;
+    res.locals.maxsize = result.rows[0].maximum || 999;
+    sqldb.query(sql.get_group_info, params, function (err, result) {
+      if (ERR(err, callback)) return;
+      res.locals.groupsize = result.rowCount;
+      res.locals.needsize = res.locals.minsize - res.locals.groupsize;
+      if (res.locals.groupsize > 0) {
+        res.locals.group_info = result.rows;
+        if (res.locals.group_info[0] === undefined) {
+          var notGroupMember = true;
+          callback(null, notGroupMember);
+        }
+        res.locals.join_code =
+          res.locals.group_info[0].name + '-' + res.locals.group_info[0].join_code;
+        res.locals.start = false;
+        if (res.locals.needsize <= 0) {
+          res.locals.start = true;
+        }
+      }
+      if (res.locals.groupsize >= res.locals.maxsize) {
+        res.locals.used_join_code = true;
+      } else {
+        res.locals.used_join_code = undefined;
+      }
+      callback();
+    });
+  });
+};
+
+module.exports.joinGroup = function (joinCode, res, callback) {
+  var splitJoinCode = joinCode.split('-');
+  const validJoinCode = splitJoinCode.length === 2;
+  if (validJoinCode) {
+    const group_name = splitJoinCode[0];
+    const join_code = splitJoinCode[1].toUpperCase();
+    if (join_code.length !== 4) {
+      throw new Error('invalid length of join code');
```

Comment:
Instead of throwing, this should be passed to `callback()`, possibly via the second argument?

Reply:
Addressed in [ecc8aee](https://github.com/PrairieLearn/PrairieLearn/pull/5290/commits/ecc8aee3e29c5c924542befe2a74e5a3d285754b). I believe since this should truly be an error that it could be passed as the first parameter which meant I could keep the 2nd callback arg as a flag, but I modified it a little for clarity (using it as `succeeded` instead of naming it like an error).